### DISCUSSION TITLE
Default settings.json

### DISCUSSION THREAD

**Discussion by I-am-PUID-0** [itsToggle/plex_debrid#268] [General]
Hi @itsToggle 

Would it be possible to add a default settings.json file to the release packages that are updated with each release? The issue I have experienced is that the format changes as new features are added. Yet, those new fields are not conveyed if utilizing a preexisting settings.json from an earlier revision unless the initial setup is performed again. This is particularly useful for docker deployments where the settings.json is provided in the /config prior to running. Ideally, it should be stored in a directory so that the volume can be mapped, thus allowing it to be pulled from the docker host and modified. 

Example of discrepancy below:

Existing settings.json:
    "Show Menu on Startup": "true",
    "Debug printing": "true",
    "version": [
        "2.20",
        "Settings compatible update",

Clean install settings.json:
    "Show Menu on Startup": "true",
    "Debug printing": "false",
    "Log to file": "false",
    "version": [
        "2.45",
        "Settings compatible update",

"Log to file" was enabled after v2.20

The same is also true for other fields, such as "Special character renaming"

THX

**Comment by itsToggle** [comment]
Hi, the plex_debrid modules take their settings from class variables, who's default values are hardcoded into each modules python file. If plex_debrid loads a settings.json file, all these class variable values are replaced with the appropriate values from the file. 
If your settings.json file does not include a new setting, the default value is used. If you save your settings at any point (Go into the settings menu and back out again), all settings (including new ones that weren't present in the old settings.json file) will be dumped into the settings.json file.
I would like to refrain from providing a default settings.json file, since the presence of such a file is used to determine if the initial setup should be run or not.

**Comment by I-am-PUID-0** [comment]
Thanks for the clarification. And admittedly, this issue may be unique to my configuration with running plex_debrid from within portainer vs. docker CLI or another platform.  

Related/unrelated: With version 2.45, when I attempt to run plex_debrid with "Show Menu on Startup": "true" or no settings.json present,  I receive the following error, and the container shuts down, which prevents entering the CLI to start the settings and exit as described. 

[v2.45]
Options/
1) Run
2) Settings
Traceback (most recent call last):
3) Ignored Media
4) Scraper
Type exit to quit.
  File "//./main.py", line 19, in <module>
    ui.run(config_dir)
  File "/ui/__init__.py", line 310, in run
    options()
  File "/ui/__init__.py", line 213, in options
    choice = input('Choose an action: ')
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
EOFError: EOF when reading a line
Choose an action: 

The above is not present when "Show Menu on Startup": "false" and the plex_debrid runs successfully. This issue is also not present in version 2.20. 

This may also be moot, as I know you're working towards a web UI version. :) 

Please also forgive my ignorance, as I don't fully understand the complexities involved. Would it be possible for plex_debrid to update the settings.json with any missing default fields upon startup, regardless of entering the settings menu? The thought is that with "Show Menu on Startup": "false" the setting.json would receive any new values available within the release and allow review/update of the fields directly from the json. Perhaps, something I could implement in CLI with the current revision? 

As for the default settings.json, would providing it as sample_settings.json mitigate the issue with the initial setup? 

Thanks again for entertaining my ideas!

**Reply by itsToggle** [reply]
That sounds like you are running the docker container without "-ti", which are the docker flags that allow for terminal input.

Yeah saving the setting after loading could easily be implemented. You would just have to call the "save()" function at the end of the "load()" function inside the "/ui/\_\_init\_\_.py" file. I could implement this in the next update :)

**Reply by itsToggle** [reply]
Despite any of this, I would really recommend utilizing the settings menu instead of editing the settings.json file directly. There are a lot of settings now which need very specific input, depend on other settings, or are oauth authenticatios which you can't input manually anyway. 
Providing anything else inside the settings.json file than what the settings menu expects could lead to unwanted behavior.

**Reply by I-am-PUID-0** [reply]
> That sounds like you are running the docker container without "-ti", which are the docker flags that allow for terminal input.

Yeah. That's because I'm using portainer -- https://hub.docker.com/r/portainer/portainer-ce -- to manage docker containers. I have plex_debrid running in a stack within portainer.

Example docker-compose .yml below: 
  Plex_Debrid:
    image: itstoggle/plex_debrid:latest
    container_name: Plex_Debrid
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/New_York
    networks:
      - proxy            
    #command: 
    volumes:
     - ./plexdebrid/config:/config

Though I can also pass commands via the commented-out command function above. 

CLI access to the container is provided through portainer, as shown below: 

![portainer 1](https://user-images.githubusercontent.com/36779668/203655139-7ee45fd6-c57f-442b-b174-08dccafc96f0.JPG)

The unfortunate consequence of my configuration is that I cannot access CLI unless the container is running, and if plex_debrid throws an error at startup, it will shut down the container. 

> Despite any of this, I would really recommend utilizing the settings menu instead of editing the settings.json file directly. There are a lot of settings now which need very specific input, depend on other settings, or are oauth authenticatios which you can't input manually anyway. Providing anything else inside the settings.json file than what the settings menu expects could lead to unwanted behavior.

For this, I historically have used the portainer CLI to edit the settings by editing the settings.json stored in the ./plexdebrid/config volume linked to the container. I have to set "Show Menu on Startup": "true" -- whereas typically it's set to "Show Menu on Startup": "false" I then restart the container and run python ./main.py --config-dir config within the portainer CLI to get to the menu. I then make the changes, pull the settings.json, revert "Show Menu on Startup": "false" and restart the container. 

For the instance mentioned previously where the container fails to run due to the "Show Menu on Startup": "true" in version 2.45, I have a python instance on windows that I run so that I can perform the initial setup or, as discussed, create the settings.json, which I then copy over to the docker ./plexdebrid/config directory. It's a bit of a workaround, but it works.

**Reply by itsToggle** [reply]
The script is exiting because the python module can not ask for input from a non-interactive terminal. You need to enable input for your docker terminal.

Have a look at the dockerfile from this github if you want to see how to enable terminal input for docker. I dont have the time to read about how to accomplish this in portainer, but im guessing it's pretty similar. 

~The top result of a quick Google search suggested using the "docker-compose run" command instead of the "docker-compose up" command to enable input. (https://stackoverflow.com/questions/62065836/why-does-a-container-running-a-console-app-simply-exits-after-starting)~ edit: this seems to reference docker and not portainer

If I find the time to write the webui in the next couple of months, this will all be redundant anyway, so I would really suggest you get the terminal input working for your portainer setup rather than me trying to make editing the settings.json file an alternative to a guided ui menu.

**Reply by itsToggle** [reply]
Ohhh is portainer simply not capable of providing an interactive shell?

**Reply by I-am-PUID-0** [reply]
Hey @itsToggle, 

To clarify a bit, portainer does support interactive shell (interactive tty); though I was referring to it as CLI (command-line interface). That is how I am modifying the settings.json. By running python ./main.py --config-dir config within the interactive shell after setting the settings.json to "Show Menu on Startup": "true" and entering the setting menu. 

Below is another view of the interactive shell within portainer after running the above cmd with "Show Menu on Startup": "false"

![portainer 2](https://user-images.githubusercontent.com/36779668/203696975-e5a6b675-7fb7-4ae0-b24d-7f2ddf669290.PNG)

I have a few instances of plex_debrid setup, a dev/prod environment type of setup. In prod, I'm running version 2.20; in dev, I'm running 2.45. When setting "Show Menu on Startup": "true" in version 2.45, I receive the previously mentioned error. Whereas with version 2.20, I do not. I have also confirmed via a windows instance running python that version 2.45 functions properly in either state true/false. So, I believe it may be limited to either docker or my specific configuration. 

Either way, I'll continue to troubleshoot and provide updates to the community. As a benefit of portainer, I can access the interactive shell and run the setup even while the container is running under the "Show Menu on Startup": "false" configuration, thus, allowing me to modify the settings.json still. 

Thanks again for the help and excellent project! I look forward to what is to come!

**Reply by itsToggle** [reply]
hm strange that v2.20 doesnt lead to that error. Ill check out the differences in the code, perhaps there is something Ive changed in the newer versions.
Im currently testing out the change I mentioned to the load() function. Plex_debrid (v2.48 experimental) should now always save the entire settings after loading them, which should allow you to more easily edit the json settings like we discussed a few posts back :)

**Reply by itsToggle** [reply]
okay the only meaningful changes to the ui since v2.20 are:
- introduction of "--config_dir some/path" definition
- checking if the provided path for settings.json exists, the path + "/settings.json" leads to a file and the filesize is greater than 0 before loading the settings.

Those changes shouldnt affect the "EOF" error you are encountering, if Im not mistaken. Ill dig a little deeper :)
